File: rdds/cache_persist.py
from pyspark import SparkContext
from sys import stdin

sc = SparkContext("local[*]", "customers")

base_rdd = sc.textFile("path")

rdd1 = base_rdd.map( lambda x: (x.split(",")[0], float(x.split(",")[2])))

rdd2 = rdd1.reduceByKey( lambda x,y: (x+y))

rdd3 = rdd2.filter( lambda x: x[1] > 5000 )

rdd4 = rdd2.map(lambda x: (x[0], x[1]*2)).persist(storageLevel="MEMORY_ONLY")
# rdd4 = rdd2.map(lambda x: (x[0], x[1]*2)).cache()

result = rdd4.collect()

for x in result:
    print(x)
    
print(rdd4.count())

stdin.readlines()   # for programing to keep running and stop when we want it

File: rdds/cache_persist.scala
import org.apache.spark.storage.StorageLevel  // this needed to be imported for storage level

object customer extends App{

    val sc = new SparkContext(local[*], "customer")
    val input = sc.textFile("c:\\Users\\pp255070\\Downloads\\customerorders.csv")

    val required = input.map( x => ((x.split(",")(0), x.split(",")(2).toFloat)))
    val aggregated = required.reduceByKey( (x,y) => x+y)
    val filterd =  aggregated.filter( x => x._2 > 5000)
    val double = filterd.map( x => x._2 *2).cache()// there is nothing like uncache, instead there is unpersist()
    //val double = filterd.map( x => x._2 *2).persist(StorageLevel.MEMORY_ONLY)
    double.collect.foreach(println)
    println(double.count)

    double.toDebugString()  // it will give lineage graph, read it from bottom

    //val sorted = aggregated.sortBy( x => x._2)
    //val result = sorted.collect       // this is not rdd as it is not distributed others are rdds as they are distributed
    //result.saveAsTextfile("path to folder")

    //val filterd =  aggregated.filter( x => x._2 > 5000)
    // val double = filterd.map( x => x._2 *2)
    //double.collect.foreach(println)
    
    // println(double.count)
}